modules/xray/shared/src/main/scala/natchez/xray/XRaySpan.scala

```diff
@@ -0,0 +1,309 @@
+      case NumberValue(n: java.lang.Integer)    => n.asJson
+      case NumberValue(n: java.lang.Long)       => n.asJson
+      case NumberValue(n: java.lang.Float)      => n.asJson
+      case NumberValue(n: java.lang.Double)     => n.asJson
+      case NumberValue(n: java.math.BigDecimal) => n.asJson
+      case NumberValue(n: java.math.BigInteger) => n.asJson
+      case NumberValue(n: BigDecimal)           => n.asJson
+      case NumberValue(n: BigInt)               => n.asJson
+      case NumberValue(n)                       => n.doubleValue.asJson
+    }
+
+  val Header = "X-Amzn-Trace-Id"
+
+  private[xray] def encodeHeader(
+      rootId: String,
+      parentId: Option[String],
+      sampled: Boolean
+  ): String = {
+    val parent = parentId.map(p => s"Parent=$p;").getOrElse("")
+    s"Root=$rootId;${parent}Sampled=${if (sampled) "1" else "0"}"
+  }
+
+  final case class XRayHeader(
+      traceId: String,
+      parentId: Option[String],
+      sampled: Boolean
+  ) {
+    def toKernel: Kernel =
+      Kernel(Map(Header -> encodeHeader(traceId, parentId, sampled)))
+  }
+
+  private[xray] def parseHeader(header: String): Option[XRayHeader] = {
+    val foo = header
+      .split(';')
+      .toList
+      .flatMap(_.split('=') match {
+        case Array(k, v) => List((k, v))
+        case _           => List.empty
+      })
+      .toMap
+
+    foo
+      .get("Root")
+      .map(traceId =>
+        XRayHeader(traceId, foo.get("Parent"), foo.get("Sampled").contains("1"))
+      )
+  }
+
+  private def randomHexString[F[_]: Functor: Random](bytes: Int): F[String] =
+    Random[F]
+      .nextBytes(bytes)
+      .map(x => BigInt(1, x).toString(16).reverse.padTo(bytes * 2, '0').reverse)
+
+  private def segmentId[F[_]: Functor: Random]: F[String] =
+    randomHexString[F](8)
+
+  private def traceId[F[_]: Applicative: Clock: Random]: F[String] =
+    (Clock[F].realTime, randomHexString[F](12)).mapN { (t, r) =>
+      s"1-${t.toSeconds.toHexString}-$r"
+    }
+
+  def fromHeader[F[_]: Concurrent: Clock: Random](
+      name: String,
+      header: XRayHeader,
+      entry: XRayEntryPoint[F]
+  ): F[XRaySpan[F]] =
+    (
+      segmentId[F],
+      Clock[F].realTime,
+      Ref[F].of(Map.empty[String, Json]),
+      Ref[F].of(List.empty[JsonObject])
+    )
+      .mapN { (sId, t, fields, children) =>
+        XRaySpan(
+          entry = entry,
+          name = name,
+          segmentId = sId,
+          xrayTraceId = header.traceId,
+          startTime = t,
+          fields = fields,
+          children = children,
+          parent = header.parentId.map(_.asLeft),
+          sampled = header.sampled
+        )
+      }
+
+  def fromKernel[F[_]: Concurrent: Clock: Random](
+      name: String,
+      kernel: Kernel,
+      entry: XRayEntryPoint[F]
+  ): F[XRaySpan[F]] =
+    kernel.toHeaders
+      .get(Header)
+      .flatMap(parseHeader)
+      .map(x => fromHeader(name, x, entry))
+      .get
```

Comment:
I think this `.get` is unsafe because it may throw an exception outside the scope of the `F[_]` effect. One option would be to do something like this:

```scala
cats.data.OptionT.fromOption[F](kernel.toHeaders.get(Header))
  .subflatMap(parseHeader)
  .semiflatMap(fromHeader(name, _, entry))
  .getOrElseF(new NoSuchElementException().raiseError)
```